### ISSUE TITLE
go1.21 support

### ISSUE THREAD

**Issue by ldez** [golangci/golangci-lint#3933]
### 2023-08-11

go1.21 is now officially fully supported since [v1.54.1](https://github.com/golangci/golangci-lint/releases/tag/v1.54.1).

### 2023-08-10

The problems with ruleguard/gocritic are fixed.
A new release will happen quickly.

### 2023-08-09

go1.21 is now officially supported since [v1.54.0](https://github.com/golangci/golangci-lint/releases/tag/v1.54.0) but there are still problems:
- [x] `ruleguard` (a "rule" inside `go-critic`) doesn't work with go1.21. https://github.com/quasilyte/go-ruleguard/issues/449
- [x] in some cases `go-critic` fails to load its built-in/embedded rules. https://github.com/go-critic/go-critic/issues/1359

---

## Is golangci-lint support go1.21?

No, and yes.

Basically, golangci-lint supports Go versions lower or equal to the Go version used to build it.

| Go version (build) | Supported Go versions (code)     | Not supported Go versions (code) |
|--------------------|----------------------------------|----------------------------------|
| go1.20             | <= go1.20 (go1.20, go1.19, etc.) | > go1.20 (go1.21, go1.22, etc.)  |
| go1.19             | <= go1.19 (go1.19, go1.18, etc.) | > go1.19 (go1.20, go1.21, etc.)  |

The current binaries from the releases page and the Docker images have been built with go1.20 and will be built with go1.20 as long as go1.21 is not GA.
So those binaries/images don't support go1.21.

https://golangci-lint.run/usage/faq/#which-go-versions-are-supported

In addition, of the Go version used to build it, some linters could need to be updated to support newer versions of Go, and internal pieces of golangci-lint could also need to be updated.

You can follow the PR https://github.com/golangci/golangci-lint/pull/3922 to see the progress on this topic.

---

_Sponsoring is a good way to sustain open source maintainers: [sponsor me](https://github.com/sponsors/ldez)_

**Comment by ldez** [comment]
[go1.21](https://go.dev/blog/go1.21) has been released.

But currently `ruleguard` (a "rule" inside `go-critic`) doesn't work with go1.21.
https://github.com/quasilyte/go-ruleguard/issues/449

We have several solutions:
- fix `ruleguard` (the best but not trivial)
- disable `ruleguard` inside `go-critic` for go1.21 (can be complex)
- disable `ruleguard` inside `golangci-lint` for go1.21 (not trivial but I think it's possible)
- disable `ruleguard` inside `golangci-lint`
- disable go-critic inside `golangci-lint` for go1.21 (the worst, I think we don't need to do that)

**Comment by andig** [comment]
In terms of a quick fix and limited-effort band-aid it sounds as if

> disable ruleguard inside golangci-lint

might be the best option right now as long a fixing the root cause is not possible?

**Comment by ldez** [comment]
I also think that is the best move: https://github.com/golangci/golangci-lint/pull/3922#issuecomment-1670243139

**Comment by cornelk** [comment]
gci needs to be upgraded by https://github.com/daixiang0/gci/pull/163 to support the new packages

**Comment by CubicrootXYZ** [comment]
go 1.21 adds a new `toolchain` directive to the `go.mod`. This currently throws an error:

```
level=error msg="Running error: context loading failed: failed to load packages: failed to load with go/packages: err: exit status 1: stderr: go: errors parsing go.mod:\n/builds/xxxxxx/go.mod:5: unknown directive: toolchain\n"
```

**Comment by ldez** [comment]
The `go.mod` structure inside `x/mod` has been updated in v0.11.0
https://github.com/golang/mod/blob/62c7e578f1a7275d934c99dd48715525bd52b17e/modfile/rule.go#L40

`x/mod` has been updated to v0.11.0 inside golangci-lint: https://github.com/golangci/golangci-lint/pull/3943

**Comment by LKaemmerling** [comment]
> The `go.mod` structure inside `x/mod` has been updated in v0.11.0 https://github.com/golang/mod/blob/62c7e578f1a7275d934c99dd48715525bd52b17e/modfile/rule.go#L40
> 
> `x/mod` has been updated to v0.11.0 inside golangci-lint: #3943

Collegue of @CubicrootXYZ here, yep this is correct, but there was no new release since then. So golangci-lint v1.53.3 was released on 15. June; the MR you refer to is from 09.July

**Comment by ldez** [comment]
> So golangci-lint v1.53.3 was released on 15. June; the MR you refer to is from 09.July

Yes, I explained that in the issue description: https://github.com/golangci/golangci-lint/issues/3933#issue-1780746540

**Comment by kikihakiem** [comment]
I have an issue with `golangci-lint` after I upgraded to go1.21. This is the command I used to install from source and run the linter: 

```sh
$ go1.21.0 install github.com/golangci/golangci-lint/cmd/golangci-lint@master
$ golangci-lint run
```

And this is the error message:

```
WARN [runner] Can't run linter goanalysis_metalinter: buildssa: failed to load package : could not load export data: no export data for "log/slog" 
ERRO Running error: 1 error occurred:
        * can't run linter goanalysis_metalinter: buildssa: failed to load package : could not load export data: no export data for "log/slog"
```

If this issue is not related to `golangci-lint`, what could be the cause of it? Could you please shed some light?

**Comment by ldez** [comment]
can you provide a reproducible example?

**Comment by exfly** [comment]
Meet new error:
`9999999999999 (untyped int constant) overflows int`

**Comment by ldez** [comment]
If you are facing a problem related to Go1.21, the best way to contribute is to open a new issue and provide the following information:

```
golangci-lint --version
```

```
cat .golangci.yml
```

```
go version && go env
```

```
golangci-lint cache clean
golangci-lint run -v
```

And a code example or link to a public repository.

**Comment by nhatthm** [comment]
New Issue:

```
❯ golangci-lint-v1.54.0 version
golangci-lint has version 1.54.0 built with go1.21.0 from c1d8c565 on 2023-08-09T11:50:00Z
```

```
❯ cat .golangci.yml
# See https://github.com/golangci/golangci-lint/blob/master/.golangci.example.yml
run:
  tests: true

linters-settings:
  errcheck:
    check-type-assertions: true
    check-blank: true
  gocyclo:
    min-complexity: 20
  dupl:
    threshold: 100
  misspell:
    locale: US
  nolintlint:
    allow-leading-space: true
  unused:
    check-exported: false
  unparam:
    check-exported: true
  wrapcheck:
    ignoreSigs:
      - .Errorf(
      - errors.New(
      - errors.Unwrap(
      - .Wrap(
      - .Wrapf(
      - .WithMessage(
      - .WithMessagef(
      - .WithStack(
      - ctxd.WrapError(
      - ctxd.NewError(
    ignorePackageGlobs:
      # Unmarshalling error is usually wrapped by the caller.
      - encoding/*

linters:
  enable-all: true
  disable:
    - exhaustivestruct
    - exhaustruct
    - forbidigo
    - gci
    - gochecknoglobals
    - golint
    - gomoddirectives
    - ireturn
    - lll
    - nonamedreturns
    - tagliatelle
    - varnamelen
    # Linters are deprecated.
    - deadcode
    - maligned
    - nosnakecase
    - scopelint
    - varcheck
    # Linters are disabled because of go1.18.
    - ifshort
    - interfacer
    - structcheck

issues:
  exclude-use-default: false
  exclude-rules:
    - linters:
        - dupl
        - funlen
        - goconst
        - goerr113
        - gomnd
        - maintidx
        - noctx
        - nosnakecase
      path: "_test.go"
```

```
❯ go version && go env
go version go1.21.0 darwin/arm64
GO111MODULE='on'
GOARCH='arm64'
GOBIN=''
GOCACHE='/path/to/Library/Caches/go-build'
GOENV='/path/to/Library/Application Support/go/env'
GOEXE=''
GOEXPERIMENT=''
GOFLAGS=''
GOHOSTARCH='arm64'
GOHOSTOS='darwin'
GOINSECURE=''
GOMODCACHE='/path/to/go/pkg/mod'
GONOPROXY='github.com/<owner>'
GONOSUMDB='github.com/<owner>'
GOOS='darwin'
GOPATH='/path/to/go'
GOPRIVATE='github.com/<owner>'
GOPROXY='https://proxy.golang.org,direct'
GOROOT='/path/to/sdk/go1.21.0'
GOSUMDB='sum.golang.org'
GOTMPDIR=''
GOTOOLCHAIN='auto'
GOTOOLDIR='/path/to/sdk/go1.21.0/pkg/tool/darwin_arm64'
GOVCS=''
GOVERSION='go1.21.0'
GCCGO='gccgo'
AR='ar'
CC='clang'
CXX='clang++'
CGO_ENABLED='1'
GOMOD='/path/to/go/src/github.com/<owner>/<repo>/go.mod'
GOWORK='/path/to/go/src/github.com/<owner>/<repo>/go.work'
CGO_CFLAGS='-O2 -g'
CGO_CPPFLAGS=''
CGO_CXXFLAGS='-O2 -g'
CGO_FFLAGS='-O2 -g'
CGO_LDFLAGS='-O2 -g'
PKG_CONFIG='pkg-config'
GOGCCFLAGS='-fPIC -arch arm64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -ffile-prefix-map=/var/folders/qf/5w661l852lzf6n_m2qy41csc0000gp/T/go-build1684254419=/tmp/go-build -gno-record-gcc-switches -fno-common'
```

```
❯ golangci-lint-v1.54.0 cache clean
❯ golangci-lint-v1.54.0 run -c /path/to/.golangci.yml -v
INFO [config_reader] Used config file /path/to/.golangci.yml
INFO [lintersdb] Active 86 linters: [asasalint asciicheck bidichk bodyclose containedctx contextcheck cyclop decorder depguard dogsled dupl dupword durationcheck errcheck errchkjson errname errorlint execinquery exhaustive exportloopref forcetypeassert funlen ginkgolinter gocheckcompilerdirectives gochecknoinits gocognit goconst gocritic gocyclo godot godox goerr113 gofmt gofumpt goheader goimports gomnd gomodguard goprintffuncname gosec gosimple gosmopolitan govet grouper importas ineffassign interfacebloat loggercheck maintidx makezero mirror misspell musttag nakedret nestif nilerr nilnil nlreturn noctx nolintlint nosprintfhostport paralleltest prealloc predeclared promlinter reassign revive rowserrcheck sqlclosecheck staticcheck stylecheck tagalign tenv testableexamples testpackage thelper tparallel unconvert unparam unused usestdlibvars wastedassign whitespace wrapcheck wsl zerologlint]
INFO [loader] Go packages loading at mode 575 (name|exports_file|deps|files|imports|types_sizes|compiled_files) took 857.502792ms
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 700.166µs
INFO [linters_context] importas settings found, but no aliases listed. List aliases under alias: key.
INFO [linters_context/goanalysis] analyzers took 1m23.914961045s with top 10 stages: buildir: 33.753623182s, buildssa: 26.457933947s, fact_deprecated: 3.720364053s, exhaustive: 3.384043725s, ctrlflow: 2.515430218s, inspect: 2.323983512s, printf: 2.185591264s, nilness: 2.068218582s, typedness: 1.588373128s, fact_purity: 1.545904843s
WARN [runner] Can't run linter goanalysis_metalinter: musttag: decoding json: invalid character '{' after top-level value: {
	"Path": "github.com/<owner>/<repo>",
	"Main": true,
	"Dir": "/path/to/go/src/github.com/<owner>/<repo>",
	"GoMod": "/path/to/go/src/github.com/<owner>/<repo>/go.mod",
	"GoVersion": "1.21"
}
INFO [runner] processing took 2.125µs with stages: skip_dirs: 417ns, max_same_issues: 417ns, nolint: 208ns, exclude-rules: 167ns, path_prettifier: 166ns, autogenerated_exclude: 125ns, max_from_linter: 83ns, identifier_marker: 83ns, skip_files: 83ns, fixer: 42ns, cgo: 42ns, diff: 42ns, exclude: 42ns, severity-rules: 42ns, filename_unadjuster: 42ns, source_code: 42ns, sort_results: 41ns, path_shortener: 41ns, uniq_by_line: 0s, max_per_file_from_linter: 0s, path_prefixer: 0s
INFO [runner] linters took 8.7769465s with stages: goanalysis_metalinter: 8.776879917s
ERRO Running error: 1 error occurred:
	* can't run linter goanalysis_metalinter: musttag: decoding json: invalid character '{' after top-level value: {
	"Path": "github.com/<owner>/<repo>",
	"Main": true,
	"Dir": "/path/to/go/src/github.com/<owner>/<repo>",
	"GoMod": "/path/to/go/src/github.com/<owner>/<repo>/go.mod",
	"GoVersion": "1.21"
}


INFO Memory: 98 samples, avg is 1031.8MB, max is 1932.9MB
INFO Execution took 9.644628209s
```

> And a code example or link to a public repository.

Private repo.

**Comment by ldez** [comment]
@nhatthm you have to open a new issue: https://github.com/golangci/golangci-lint/issues/new/choose

**Comment by ldez** [comment]
go1.21 is now officially fully supported since [v1.54.1](https://github.com/golangci/golangci-lint/releases/tag/v1.54.1).

---

_Sponsoring is a good way to sustain open source maintainers: [sponsor me](https://github.com/sponsors/ldez)_

**Comment by ramseyjiang** [comment]
<img width="1132" alt="Screenshot 2023-12-07 at 19 04 48" src="https://github.com/golangci/golangci-lint/assets/6574141/5891d05f-ff7f-4cbc-b52a-f7dc0d0afc3d">
Error is still here, not fixed golangci-lint 1.55.2
